src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -611,6 +621,50 @@ private static TokenCredentialData CreateTokenCredentialInstance(TokenCredential
 
                 return new TokenCredentialData(new WorkloadIdentityCredential(options), GetHash(secret));
             }
+            else if (tokenCredentialKey._tokenCredentialType == typeof(ClientAssertionCredential))
+            {
+                // We'll keep using common environment variables defined in Azure.Identity library
+                // AZURE_TENANT_ID: The tenant ID where target resource hosts
+                // AZURE_CLIENT_ID: The client id of multi-tenant entra id app. May be overridden by the User Id.
+                // AZURE_MSI_CLIENT_ID (new): The client id of managed identity
+                string tenantId = Environment.GetEnvironmentVariable("AZURE_TENANT_ID");
+                tenantId = string.IsNullOrWhiteSpace(tenantId) ? null : tenantId;
+
+                string clientId = tokenCredentialKey._clientId;
+                if (clientId is null)
+                {
+                    clientId = Environment.GetEnvironmentVariable("AZURE_CLIENT_ID");
+                    clientId = string.IsNullOrWhiteSpace(clientId) ? null : clientId;
+                }
+
+                string managedIdentityClientId = Environment.GetEnvironmentVariable("AZURE_MSI_CLIENT_ID");
+                managedIdentityClientId = string.IsNullOrWhiteSpace(managedIdentityClientId) ? null : managedIdentityClientId;
+
+                ClientAssertionCredential clientAssertionCredential = null;
+                // If either tenant id, client id, or the managed identity client id are not specified when fetching the token,
+                // a CredentialUnavailableException will be thrown instead
+                if (!string.IsNullOrEmpty(tenantId) && !string.IsNullOrEmpty(clientId) && !string.IsNullOrEmpty(managedIdentityClientId))
+                {
+                    clientAssertionCredential = new ClientAssertionCredential(
+                        tenantId,
+                        clientId,
+                        async (cancellationToken) =>
+                        {
+                            ManagedIdentityCredential miCredential = new ManagedIdentityCredential(managedIdentityClientId);
+                            AccessToken token = await miCredential.GetTokenAsync(
+                                new TokenRequestContext(new string[] { $"api://AzureADTokenExchange/.default" }),
+                                cancellationToken).ConfigureAwait(false);
+                            return token.Token;
```

Comment:
We should be able to cache and re-use a Managed Identity credential here, too. Something like this, I think (definitely double check this code):
```suggestion
                            string scope = "api://AzureADTokenExchange/.default";
                            TokenRequestContext tokenRequestContext = new TokenRequestContext(new string[] { scope });
                            TokenCredentialKey tokenCredentialKey = new(typeof(ManagedIdentityCredential), tokenCredentialKey._authority, scope, string.Empty, managedIdentityClientId);
                            AccessToken token = await GetTokenAsync(
                                tokenCredentialKey,
                                string.Empty,
                                tokenRequestContext,
                                cancellationToken).ConfigureAwait(false);
                            return token.Token;
```

Reply:
Updated